students/luyao_xu/lesson09/mailroom5.py

```diff
@@ -0,0 +1,249 @@
+            f' Have a nice day!'
+
+    def add_donations(self, donation_amount):
+        """
+        Add a donation.
+        :param donation_amount: Donation amount
+        :return: Donor object
+        """
+        self.donations.append(donation_amount)
+        return self
+
+
+class Database(object):
+    """
+    A collection of donor objects that holds summarized information about the fund.
+    """
+    def __init__(self, donors=None):
+        """
+        Initializes a database.
+        :param donors: List of donors to initialize the database with.
+        """
+        if donors is None:
+            self.donors = []
+        else:
+            self.donors = donors
+
+    def add_donor(self, donor):
+        """
+        Add a donor to the collection.
+        :param donor: Donor object to be added.
+        :return: None
+        """
+        if isinstance(donor, Donor):
+            self.donors.append(donor)
+        else:
+            print('Please add a valid donor.')
+
+    @property
+    def all_names(self):
+        return [d.name for d in self.donors]
+
+    def create_report(self):
+        """
+        Create a tabular report of the current donation status.
+        :return: A formatted string of the table.
+        """
+        result = '{:<26} | {:^11} | {:^8} | {:>12}\n'.format('Donor Name', 'Total Given', 'Num Gifts', 'Average Gift')
+        result += '-' * len(result) + '\n'
+        for d in self.donors:
+            name = d.name
+            total_donation = d.total_donation
+            number_of_gifts = d.number_of_gifts
+            average_donation = d.average_donation
+
+            result += '{:<27} ${:>11.2f} {:>12} ${:>13.2f}\n'.format(name, total_donation, number_of_gifts,
+                                                                     average_donation)
+
+        return result
+
+    def thank_all(self, folder='letters/'):
+        """
+        Outputs files with grateful notes for each donor.
+        :param folder: Output folder of the files.
+        :return: None
+        """
+        for donor in self.donors:
+            filename = Path(f'./{folder}{donor.name}.txt')
+            filename.parent.mkdir(parents=True, exist_ok=True)
+            with filename.open('w') as outfile:
+                outfile.write(donor.thank_you_note)
+
+
+if __name__ == '__main__':
+    def initialize_donors():
+        """
+        Initialize default donors.
+        :return: List of default donors
+        """
+        don_1 = Donor('Amy Walker', [18900.90, 4500])
+        don_2 = Donor('Peter Thomson', [9999.99, 500, 6000.88])
+        don_3 = Donor('July Jensen', [5300.00, 200])
+        don_4 = Donor('Paul Allen', [320.57])
+        don_5 = Donor('Jenny Palmer', [66.89])
+        return [don_1, don_2, don_3, don_4, don_5]
+
+    db = Database(initialize_donors())
+
+
+    def donate(donor):
```

Comment:
these functions (including ones below) should be defined outside the __name__ == __main__ block (right above it).